[ROCM] Use hipblaslt version from hipblaslt runtime instead of header for tunableops validator

zixi-qi opened the pull request:
Summary:
When tunable ops load selected kernels from csv file, it will validate hipblaslt version defined in hipblaslt-version.h

This PR changes the validator to fetch hipblaslt version and revision from hipblaslt runtime instead of the header file, as in our environment we might rollout a new version of the run time prior to updating the header file fleet wide.

Test Plan:
Verified generated tunableops kernel selection has the correct hipblaslt version from runtime:

```
Validator,PT_VERSION,2.5.0
Validator,ROCBLAS_VERSION,4.0.0-72e57364-dirty
Validator,HIPBLASLT_VERSION,800-bf2c3184
Validator,ROCM_VERSION,6.0.0.0-12969-1544e39
Validator,GCN_ARCH_NAME,gfx942:sramecc+:xnack-
GemmTunableOp_BFloat16_TN,tn_8192_2_3584,Gemm_Hipblaslt_TN_572,0.0240676
GemmTunableOp_BFloat16_TN,tn_7168_2_8192,Gemm_Hipblaslt_TN_482,0.0359019
GemmTunableOp_BFloat16_TN,tn_8192_2_1024,Default,0.0173723
GemmTunableOp_BFloat16_TN,tn_1280_2_8192,Gemm_Hipblaslt_TN_491,0.0191047
```

Differential Revision: D59889043


cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

jeffdaily commented on aten/src/ATen/cuda/tunable/TunableGemm.h:
```diff
@@ -194,15 +203,21 @@ static void AddRocblasValidator() {
-        XSTRINGIFY(HIPBLASLT_VERSION_PATCH), "-",
-        XSTRINGIFY(HIPBLASLT_VERSION_TWEAK));
+    int version;
+    std::string revision = "";  
```
```suggestion
    std::string revision(128, '\0');
```
Recommend either calling `revision.reserve(128)` or the above suggestion of using fill constructor with 128 null bytes.  This matches the [client example in the hipblaslt sources](https://github.com/ROCm/hipBLASLt/blob/b5c62e6fc2846b878277e1a08b30d5c6bb672161/clients/benchmarks/client.cpp#L255C5-L255C44).  The [hipblasltGetGetRivision() ](https://github.com/ROCm/hipBLASLt/blob/b5c62e6fc2846b878277e1a08b30d5c6bb672161/library/src/amd_detail/hipblaslt.cpp#L674)will do a memcpy to the buffer passed in and you haven't allocated any space using the default string constructor.

zixi-qi replied:
@jeffdaily Thank you for the guidance! Updated the PR accordingly

jeffdaily commented on aten/src/ATen/cuda/tunable/TunableGemm.h:
```diff
@@ -194,15 +203,21 @@ static void AddRocblasValidator() {
-        XSTRINGIFY(HIPBLASLT_VERSION_TWEAK));
+    int version;
+    std::string revision = "";  
+    hipblasLtHandle_t handle;
+    TORCH_HIPBLASLT_CHECK(hipblasLtCreate(&handle));
+    hipblasLtGetVersion(handle, &version);
+    hipblasLtGetGitRevision(handle, revision.data());
+    TORCH_HIPBLASLT_CHECK(hipblasLtDestroy(handle));
```
```suggestion
    auto handle = at::cuda::getCurrentCUDABlasLtHandle();
    hipblasLtGetVersion(handle, &version);
    hipblasLtGetGitRevision(handle, revision.data());
```
Using current hipblaslt handle means you don't have to free it since it is pooled, and you can remove the `TORCH_HIPBLASLT_CHECK` added above.

jeffdaily commented on aten/src/ATen/cuda/tunable/TunableGemm.h:
```diff
@@ -25,6 +25,15 @@
 
 #ifdef USE_ROCM
 #include <rocm-core/rocm_version.h>
+#define TORCH_HIPBLASLT_CHECK(EXPR)      \
+  do {                                   \
+    hipblasStatus_t __err = EXPR;        \
+    TORCH_CHECK(                         \
+        __err == HIPBLAS_STATUS_SUCCESS, \
+        "hipblaslt error: ",             \
+        hipblasStatusToString(__err),    \
+        " when calling `" #EXPR "`");    \
+  } while (0)
```
```suggestion
```
Don't need, see below.

jeffdaily reviewed (changes requested):
Conceptually the PR makes sense -- use the runtime-queried version instead of the header.  I'm good with that.  I'm concerned about
- the `std::string revision` not allocating space to hold the result.
- use hipblaslt handle from pool to simplify the code

xw285cornell reviewed (approved):
nice work!

xw285cornell commented:
I wonder if we can add some unit test for this. E.g. mock the hipblaslt get version call.

zixi-qi commented:
> I wonder if we can add some unit test for this. E.g. mock the hipblaslt get version call.

Added a basic unit test to validate the hipblaslt version validator is created and has the expected format in python unit test. CPP side would require more setup work, can do that as a follow up work

Commit:
[AMD] Add hipblaslt runtime version to tunableops validators (#131078)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/131078

When tunable ops load selected kernels from csv file, it will only validate hipblaslt version defined in hipblaslt-version.h: https://fburl.com/code/i9khty4g

However in our current hipblaslt runtime rollout workflow, we might rollout libhipblaslt.so prior to updating hipblaslt-version.h fleet-wide, which would cause a discrepancy between the hipblaslt version used in tunableops validation and the actual runtime hipblaslt version. This diff adds a new validation to tunable ops by checking the runtime hipblaslt version in addition to the version defined in hipblaslt-version.h so that we can cover the mismatch case in validation.

Test Plan:
Verified new validator shows up and kernel selection can be successfully saved and loaded (**Validator HIPBLASLT_RUNTIME_VERSION=800-a15e4178**):
```
[qizixi@devgpu038.atn3 /data/users/qizixi/fbsource/fbcode (493e70f56)]$ buck2 run mode/{opt,amd-gpu} -c fbcode.enable_gpu_sections=true //scripts/xdwang/example:fc_llama -- --enable-tuning
Buck UI: https://www.internalfb.com/buck2/215d6dcb-8c21-473b-b6fe-19c3b0753708
Network: Up: 0B  Down: 0B
Jobs completed: 4531. Time elapsed: 0.6s.
BUILD SUCCEEDED
Enabled tuning
- Run Linear (matmul) 2 x 1280 x 8192, dtype = torch.bfloat16
INFO:2024-07-17 17:01:18 1884589:1884986 CuptiActivityProfiler.cpp:260] HIP versions. Roctracer: 4.1; Runtime: 60032830; Driver: 60032830
INFO:2024-07-17 17:01:18 1884589:1885495 DynoConfigLoader.cpp:61] Setting communication fabric enabled = 0
reading tuning results from hipblas_tuning_pt_llama0.csv
Validator PT_VERSION=2.5.0
Validator ROCBLAS_VERSION=4.0.0-72e57364-dirty
Validator HIPBLASLT_VERSION=0.8.0-a15e4178
Validator ROCM_VERSION=6.0.0.0-12969-1544e39
Validator HIPBLASLT_RUNTIME_VERSION=800-a15e4178
Validator GCN_ARCH_NAME=gfx942:sramecc+:xnack-
Loading results
GemmTunableOp_BFloat16_TN(tn_1280_2_8192) -> Gemm_Hipblaslt_TN_61098,0.0265708
GemmTunableOp_BFloat16_TN(tn_8192_2_1024) -> Gemm_Hipblaslt_TN_61089,0.0250619
GemmTunableOp_BFloat16_TN(tn_7168_2_8192) -> Gemm_Hipblaslt_TN_61089,0.0466506
GemmTunableOp_BFloat16_TN(tn_8192_2_3584) -> Gemm_Hipblaslt_TN_61089,0.0363046
Avg time: 13.150180578231812 us, Achieved 3.19 TFLOPS, 1600.14 GB/s

- Run Linear (matmul) 2 x 8192 x 1024, dtype = torch.bfloat16
Avg time: 13.361680507659912 us, Achieved 2.51 TFLOPS, 1258.69 GB/s

- Run Linear (matmul) 2 x 7168 x 8192, dtype = torch.bfloat16
Avg time: 27.896180152893066 us, Achieved 8.42 TFLOPS, 4213.29 GB/s

- Run Linear (matmul) 2 x 8192 x 3584, dtype = torch.bfloat16
Avg time: 17.520229816436768 us, Achieved 6.70 TFLOPS, 3355.08 GB/s

2x1280x8192-torch.bfloat16,13.150180578231812,3.1895409915077915,1600.1435018186917
2x8192x1024-torch.bfloat16,13.361680507659912,2.511243400915334,1258.6871831247997
2x7168x8192-torch.bfloat16,27.896180152893066,8.419827471455473,4213.290828916972
2x8192x3584-torch.bfloat16,17.520229816436768,6.7031376431958725,3355.0756249129454
WARNING:2024-07-17 17:01:38 1884589:1885495 DynoConfigLoader.cpp:35] (x1) Failed to read config: No dyno config client
```

Reviewed By: xw285cornell

Differential Revision: D59889043

clee2000 commented:
@pytorchbot merge -f "merged internally"
